File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveMIMETypeInputName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    let parameter = getParent(node)
    if (ts.isUnionTypeNode(parameter)) {
        parameter = getParent(parameter)
    }

    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null

    const parameterName = parameter.name.text
    if (parameterName !== "input") return null

    const constructor = getParent(parameter)
    if (!constructor) return null
    if (!ts.isConstructorDeclaration(constructor)) return null

    const classNode = getParent(constructor)
    if (!classNode) return null
    if (!ts.isClassDeclaration(classNode)) return null
    if (classNode.name === undefined) return null

    const parentName = classNode.name.text
    if (parentName !== "MIMEType") return null

    return `${karakum.capitalize(parentName)}${karakum.capitalize(parameterName)}`
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/nameResolvers/resolveMIMETypeInputName.kt
package wrappersgenerator.node.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.*

fun resolveMIMETypeInputName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

    val parameter = ensureNotNull(
        typeScriptService.getParent(node)
            ?.let {
                if (isUnionTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text
    ensure(parameterName == "input")

    val constructor = ensureNotNull(typeScriptService.getParent(parameter))
    ensure(isConstructorDeclaration(constructor))

    val classNode = ensureNotNull(typeScriptService.getParent(constructor))
    ensure(isClassDeclaration(classNode))
    val parentName = ensureNotNull(classNode.name).text
    ensure(parentName == "MIMEType")

    "${parentName.replaceFirstChar { it.titlecase() }}${parameterName.replaceFirstChar { it.titlecase() }}"
}
